src/middlewared/middlewared/plugins/acm.py

```diff
@@ -0,0 +1,736 @@
+            )
+
+        if self.middleware.call_sync('acme.registration.query', [['directory', '=', data['directory_uri']]]):
+            verrors.add(
+                'acme_registration_create.directory',
+                'A registration with the specified directory uri already exists'
+            )
+
+        if verrors:
+            raise verrors
+
+        key = jose.JWKRSA(key=rsa.generate_private_key(
+            public_exponent=data['JWK_create']['public_exponent'],
+            key_size=data['JWK_create']['key_size'],
+            backend=default_backend()
+        ))
+        acme_client = client.ClientV2(directory, client.ClientNetwork(key))
+        register = acme_client.new_account(
+            messages.NewRegistration.from_data(
+                email=email,
+                terms_of_service_agreed=True
+            )
+        )
+        # We have registered with the acme server
+
+        # Save registration object
+        registration_id = self.middleware.call_sync(
+            'datastore.insert',
+            self._config.datastore,
+            {
+                'uri': register.uri,
+                'tos': register.terms_of_service,
+                'new_account_uri': directory.newAccount,
+                'new_nonce_uri': directory.newNonce,
+                'new_order_uri': directory.newOrder,
+                'revoke_cert_uri': directory.revokeCert,
+                'directory': data['directory_uri'] + '/' if data['directory_uri'][-1] != '/' else ''
+            }
+        )
+
+        # Save registration body
+        self.middleware.call_sync(
+            'datastore.insert',
+            'system.acmeregistrationbody',
+            {
+                'contact': register.body.contact[0],
+                'status': register.body.status,
+                'key': key.json_dumps(),
+                'acme': registration_id
+            },
+        )
+
+        return self.middleware.call_sync(f'{self._config.namespace}.query', [('id', '=', registration_id)])[0]
+
+
+class ACMEService(CRUDService):
+
+    class Config:
+        datastore = 'system.certificate'
+        datastore_prefix = 'cert_'
+        datastore_extend = 'certificate.cert_extend'
+        private = True
+
+    async def query(self, filters=None, options=None):
+        if not filters:
+            filters = [['acme', '!=', None]]
+        else:
+            filters.append(['acme', '!=', None])
+        return await super(ACMEService, self).query(filters, options)
+
+    @private
+    async def get_domain_names(self, data):
+        names = [data['common']]
+        names.extend(data['san'])
+        return names
+
+    @accepts(
+        Dict(
+            'acme_create',
+            Bool('tos', default=False),
+            Int('csr_id', required=True),
+            Str('directory_uri', required=True),
+            Str('name', required=True),
+            Dict('domain_dns_mapping', additional_attrs=True, required=True)
```

Comment:
`domain_dns_mapping` that name seem a bit redundant, since DNS stands for domain name server